ValueObject/ValueObject.cs

```diff
@@ -55,24 +53,26 @@ private bool FieldsAreEqual(object obj, FieldInfo f)
 
         private IEnumerable<PropertyInfo> GetProperties()
         {
-            if (this.properties == null)
+            Type type = GetType();
+            if (!properties.ContainsKey(type))
             {
-                this.properties = GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                properties[type] = GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public)
                     .Where(p => !Attribute.IsDefined(p, typeof(IgnoreMemberAttribute))).ToList();
             }
 
-            return this.properties;
+            return properties[type];
         }
 
         private IEnumerable<FieldInfo> GetFields()
         {
-            if (this.fields == null)
+            Type type = GetType();
+            if (!fields.ContainsKey(type))
             {
-                this.fields = GetType().GetFields(BindingFlags.Instance | BindingFlags.Public)
+                fields[type] = GetType().GetFields(BindingFlags.Instance | BindingFlags.Public)

```

Comment:
since Dictionary is not thread safe, multiple threads can insert the same type here and break the dictionary